### ISSUE TITLE
Web: cannot use Fira Code in editor (Safari)

### ISSUE THREAD

**Issue by bpasero** [microsoft/vscode#83294]
Open web in Safari with `editor.fontFamily: "Fira Code"`. the font is not being loaded it seems. Works fine in other browsers.

**Comment by bpasero** [comment]
//cc @alexandrudima who I think knows why this happens

Maybe one workaround would be to at least not fallback to a serif-font when this happens but rather pick our default (if we can possibly detect that).

**Comment by alexandrudima** [comment]
My theory is that this is blocked to prevent finger-printing via installed fonts...

I am not sure how the editor itself can validate this setting, because the editor can be embedded in pages which do contain Fira Code via CSS...

So I think we might need special editor.fontFamily Safari validation, but only in the workbench (not the standalone editor)...

**Comment by alexandrudima** [comment]
AFAIK that piece of code only executes on startup (to load the previous font metrics from local storage and avoid measuring the font and doing a sync layout in the startup phase). It is possible to change `editor.fontFamily` after the workbench startup, so that place would not be good.

I think we need an editor contribution that only executes on the web, on Safari, and which listens to editor option change events, and checks if the `fontFamily` is configured with a font that work on Safari or not... But how should we know the entire list of fonts that work on Safari?

**Comment by alexandrudima** [comment]
I have found this -- https://en.wikipedia.org/wiki/List_of_typefaces_included_with_macOS , should we ship with this list hard coded somewhere?

**Comment by bpasero** [comment]
I like that. We can possibly put this somewhere into our workbench specific overrides for the editor? E.g. when running in Safari, we do not allow a font family that is not in that set?

Code pointer: [`getConfigurationOverrides`](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/browser/parts/editor/textEditor.ts#L121)

**Comment by alexandrudima** [comment]
@bpasero When you played with the CSS variant, did you experiment with using `local` as a `src`?

https://stackoverflow.com/a/43502304

e.g.
```
@font-face {
  font-family: 'Montserrat';
  font-style: normal;
  font-weight: 400;
  src: url("font/montserrat/montserrat-latin-regular.eot");
  src: local("Montserrat-Regular"), url("font/montserrat/montserrat-latin-regular.eot?#iefix") format("embedded-opentype"), url("font/montserrat/montserrat-latin-regular.woff2") format("woff2"), url("font/montserrat/montserrat-latin-regular.woff") format("woff"), url("font/montserrat/montserrat-latin-regular.ttf") format("truetype"), url("font/montserrat/montserrat-latin-regular.svg#Montserrat") format("svg"); 
}
```

**Comment by alexandrudima** [comment]
Actually it looks like this is a known limitation -- https://help.figma.com/article/356-use-local-fonts-with-figma-font-helper

![image](https://user-images.githubusercontent.com/5047891/67853600-34c64100-fb0f-11e9-936b-d26ec08c0d97.png)

**Comment by alexandrudima** [comment]
Yeah, I finally could find a definitive answer -- it is done to avoid fingerprinting -- https://stackoverflow.com/a/52805055

**Comment by bpasero** [comment]
I actually think this is a fair workaround:

`"editor.fontFamily": "Fira Code, Menlo, Monaco"`

It works nicely on Safari to fallback to Menlo.

**Comment by Mara-Li** [comment]
Hello! Sorry to bother. I want to know if there is a way to import font (embedded using base64, for example) to change the font in safari ? Using HTML `<style>` or a CSS file ?

Thank you :)!

**Comment by oliverlabs** [comment]
Has anyone been able to work around this? Any Nerd Font in Safari seems to be blocked.

**Comment by 7ombie** [comment]
Safari deliberately prevents workarounds (which is overkill). They removed the options a while ago (see [this StackOverflow question][soq]).

The only way to use Nerd Fonts in Safari is to serve them along with the client. I'm pretty sure browsers only load what they need (font-wise), so that would be an option, but requires changes on the server.

As a user, you'd need to achieve the same effect without access to the server (assuming you're not hosting VSCode yourself). It's may be possible to use a browser extension to inject the font into the page, or just create a browser extension for injecting the font. IIUC, as long as the font is loaded by the page, referencing it will work.

[soq]: https://apple.stackexchange.com/questions/106083/is-there-a-way-to-set-the-default-monospace-font-in-safari

**Comment by 7ombie** [comment]
Another option is to install VSCode locally and attach to a remote server:

<img width="352" alt="Screenshot 2024-02-12 at 11 43 50 PM" src="https://github.com/microsoft/vscode/assets/80077603/32472eb1-8aae-4fba-8009-b8b9bccc7b77">

This is a good choice, if you have the option.
